src/doc.cpp

```diff
@@ -1596,4 +1598,172 @@ int Doc::PrepareTimestampsEnd(FunctorParams *functorParams)
     return FUNCTOR_CONTINUE;
 }
 
+xsdAnyURI_List Doc::useExpansion(xsdAnyURI_List expansionList, xsdAnyURI_List existingList, Object *prevSect)
+{
+    for (std::string s : expansionList) {
+        if (s.rfind("#", 0) == 0) s = s.substr(1, s.size() - 1); // remove trailing hash from reference
+        Object *currSect = this->FindChildByUuid(s); // find section pointer of reference string
+        if (currSect->Is(EXPANSION)) { // if reference is itself an expansion, resolve it recursively
+            existingList = useExpansion(dynamic_cast<Expansion *>(currSect)->GetPlist(), existingList, prevSect);
+        }
+        else {
+            if (std::find(existingList.begin(), existingList.end(), s)
+                != existingList.end()) { // section exists in list
+
+                // clone current section/ending/rdg/lem and rename it, adding -"rend2" for the first repetition etc.
+                Object *clonedObject = currSect->Clone();
+                clonedObject->CloneReset();
+                clonedObject->SetUuid(currSect->GetUuid() + "-rend"
+                    + std::to_string(getExpansionIdsForElement(currSect->GetChild(0)->GetUuid()).size() + 1));
+
+                // get IDs of old and new sections and add them to m_expansionMap
+                std::vector<std::string> oldIds;
+                this->getUuidList(currSect, &oldIds);
+                std::vector<std::string> clonedIds;
+                this->getUuidList(clonedObject, &clonedIds);
+                for (int i = 0; i < oldIds.size(); i++) {
+                    this->addExpandedIdToExpansionMap(oldIds[i], clonedIds[i]);
+                }
+
+                // go through cloned objects, find TimePointing/SpanningInterface, PListInterface, LinkingInterface
+                updateIds(clonedObject);
+
+                prevSect->GetParent()->InsertAfter(prevSect, clonedObject);
+                prevSect = clonedObject;
+            }
+            else { // add to existingList, remember previous element, but do nothing else
+                prevSect = currSect;
+                existingList.push_back(s);
+            }
+        }
+    }
+    return existingList;
+}
+
+std::vector<std::string> *Doc::getUuidList(Object *object, std::vector<std::string> *idList)
+{
+    for (Object *o : *object->GetChildren()) {
+        idList->push_back(o->GetUuid());
+        idList = getUuidList(o, idList);
+    }
+    return idList;
+}
+
+bool Doc::updateIds(Object *object)
```

Comment:
This should be done through Functors. It is a bit cumbersome with interfaces but basically have one `Object::UpdateIds(FunctorParams *params)` functor methods, that would cast them (as you do here) but then call the `InterfaceUpdateIds(FunctorParams *functorParams, Object *object)` method to be defined for the InterfaceTimePoint, InterfaceTimeSpanning, etc. classes. (Look at ResetDrawing and InterfaceResetDrawing for an example).

(The loop inside the children is done by the functor mechanism)